#6 사용자로 부터 입력받고 검증하는 작업

icraft2170 opened the pull request:
사용자로 부터 입력받고 검증하는 작업
---
1. 문자열로 세자리 숫자를 입력받음
2. 입력 받는 문자열은 포맷을 지켜야함 
  - Format : nnn ( n은 10 미만의 자연수 )
3. 5번 이상 잘못 입력 시 앱이 종료되도록 함.

### 의존성 추가
- Mockito 추가되었습니다.

### 이슈
- Test 과정에 final class Mock 객체 생성 실패 이슈
- 입력 받은 값을 검증하는 로직이 조건에 부합하지 않음.

icraft2170 committed:
refactoring : #6 폴더에 클래스 분류
### 추가
- domain
  - InputNumbers
  - Numbers
  - TargetNumbers
  - PitchingResult

### 이유
- 각 클래스 명에 따라 용도를 오해할 수 있을 것으로 판단하여 이동 시켰습니다.

icraft2170 committed:
feature : #6 NumbersReader Interface 및 구현체

### <<interface>>
- NumbersReader : 입력을 받고 InputNumbers 로 반환해 주는 역할

### class ( 구현체 2종 )
- IntegerArrayReader
- StringNumbersReader

icraft2170 committed:
dependency : #6 Mockito Framework
1. Mockito 프레임워크 의존성 추가

icraft2170 committed:
issue : #6 이슈 발생 기록
- Scanner 를 Mock 객체로 사용하려고 하였으나, Scanner가 final class라 사용 불가
  -> 상속 불가가 원인으로 판단됨.

icraft2170 committed:
resolution : #6 Scanner Mock Test 관련 이슈 해결

### 참고 자료
- https://www.baeldung.com/mockito-final

### 해결 내용
- final class의 Mocking이 가능하도록 변경
- Reflection 을 활용하여 Mock Object 주입

icraft2170 committed:
test : #6 테스트 추가

### 추가된 테스트
- 중복된 숫자가 입력되었을 때
- 입력 포맷에 맞지 않은 입력이 발생했을 시
- 5번 이상 입력이 실패했을 때

icraft2170 committed:
feature : #6 문자열 입력 기능 추가

### 기능 명세
- 5회 이상 입력 실패시 `WrongInputRepeatException` 예외를 반환 함.
- 문자열을 입력 받는다.
- 입력받은 문자가 3자의 각각이 0~9까지 숫자 형식인지 확인
- 입력 포맷이 잘못되었을 때 5회까지 입력을 받는다.

### 이슈
- 중복 숫자를 검증해야하는지 고민
   - 중복 숫자를 입력하는 것이 게임의 전략으로 인정할지는 다시 한 번 생각해봐야 할 문제로 인식됩니다.

icraft2170 committed:
feat : #6 Test 수정 및 NumberBaseBallGame 에서 호출 가능하게 변경

### verifyNumberOfInput 변경
- 4회에 앱이 종료되도록 되어 있었습니다.

### 테스트 수정 이유
- 위 오류 내용을 테스트로 검증해내지 못했었다. 따라서 해당 내용을 검증할 수 있도록 함

ahngj96 commented on app/src/main/java/edu/number/baseball/util/valid/InputValidator.java:
```diff
@@ -0,0 +1,19 @@
+        char[] chars = inputNumber.toCharArray();
+
+        for (char ch : chars) {
+            isDigit = Character.isDigit(ch);
```
마지막 char값만 digit인지만 판단하는 코드인 것같습니다 !

f-lab-elkein replied:
@ahngj96 @icraft2170 

저도 그런거 같네요. 
위와 같은 버그가 발생시, 버그에 대한 확인을 하는 유닛 테스트를 추가 하면 좋습니다.

유닛 테스트 케이스중에서 첫번째 글자나 두번째 글자가 숫자가 아니고, 마지막 글자만 숫자일 경우에 대한 테스트 케이스가 추가되면 좋겠죠?

기본적으로 오동작이나 버그가 있었을 때, 수정 되었음을 증명해주는 유닛 테스트가 추가 되면 좋습니다.

icraft2170 replied:
해당 내용 다시 확인해서 체크 하겠습니다
감사해요! 😀😀

ahngj96 commented:
디렉토리 정리 감사합니다!

궁금한 점이 있다면 domain이라는 디렉토리 이름에 대해서입니다!
domain 이름의 디렉토리는 포괄하는 의미가 크거나 사람마다 받아들이는 의미도 조금씩 다르기도 해서요

전에 프로젝트에서 저는 domain이라는 디렉토리 이름을 사용하는 대신에 db entity들을 가지고 있는 entity 디렉토리와 비즈니스 로직 사이에서 다루는 불변 객체 vo디렉토리를 따로 명시적으로 분리해서 정리했었습니다
현호님은 원래 어떻게 정리하시나요?

나머지 util과 validation, exception은 비슷하게 정리하는데 domain이라는 디렉토리 이름을 사용하시기에 질문드립니다!

ahngj96 commented on app/src/main/java/edu/number/baseball/util/io/IntegerArrayReader.java:
```diff
@@ -0,0 +1,19 @@
+
+        return null;
+    }
+}
```
앜ㅋㅋㅋㅋ 제가 인풋넘버 객체에 integer list input 생성자를 만들어서 만드신거군요

안만드셔도 됐는데 ㅠㅠ 고생하셨습니다

icraft2170 replied:
아닙니다!  😀😀 여러 형태로 대응할 수 있게 배려해주셔서 감사합니다!

ahngj96 commented on app/src/test/java/edu/number/baseball/util/io/NumbersReaderTest.java:
```diff
@@ -0,0 +1,139 @@
+     * 1. 중복 숫자를 입력하여 추리하는 것이 게임의 전략으로서 인정받을 수 있는 가?
+     * 2. 1번이 인정될 때 중복 숫자 포함은 검증 목록에서 제외되어야 함.
+     */
+
```
저는 전략으로 인정할 수 있을 것같습니다!

icraft2170 replied:
저도 하다보니 생각이 들어서ㅎㅎ😀😀😀

ahngj96 commented on app/src/test/java/edu/number/baseball/util/io/NumbersReaderTest.java:
```diff
@@ -0,0 +1,139 @@
+        reflectionReader.setAccessible(true);
+        reflectionReader.set(reader, mockScanner);
+    }
+}
```
스캐너 mocking을 해본적이 없어 처음보는데 배워갑니다!

icraft2170 replied:
저도 어떻게 해야하지 하다가 추가 해보긴 했는데.. 이렇게 하는게 맞는지는 잘 모르겠어요 ㅠㅠ

icraft2170 commented:
넵!! 저는 보통 domain에 VO 혹은 Entity를 넣기는 합니다. 
개인적인 이유로는 VO에 비즈니스 로직이 포함될 수 있기 때문에 도메인 레이어에 포함될 수 있다고 판단하기 때문입니다. 😀😀
그 때문에 VO와 Entity를  도메인 패키지에 포함 시키지만 DTO는 분리해 넣는 편입니다.

다만, 요즘은 각 도메인 별로 구분하고 내부에 util, service와 같은 패키지를 도메인 별로 따로 위치 시키고 있습니다.

위 형태는 나중에 협의해서 프로젝트 기간에 결정해보면 좋을것 같습니다!!

icraft2170 commented:
건주님 혹시 머지 해도 괜찮을까요? 🤗🤗

ahngj96 commented:
제가 맨 위에 코멘트한 digit인지 체크하는 코드 문제 없는건가요?

제가 보기에 for loop전에 digit ture로 하고
루프 안에서 만약 digit이 아니면 false로 대입하고 break 걸어야할 것같아서요

이부분 해결되면 바로 머지하시면 될것같아요 👍

ahngj96 commented:
아 죄송합니다 제가 submit을 안눌렀네요 ㅠㅠㅠ

icraft2170 committed:
Test : #6 InputValidator에 static Method에 대한 단위 테스트 작성

1. 타겟에 문자열 길이가 3이며 각 자리수가 10 이하의 자연수로 구성된 정상 문자열인 경우
2. 타겟에 문자열 길이가 3이지만, 첫 자리가 조건에 적합하지 않은경우
3. 타겟에 문자열 길이가 3이지만, 중간 자리가 조건에 적합하지 않은경우
4. 타겟에 문자열 길이가 3이지만, 마지막 자리가 조건에 적합하지 않은경우
4. 타겟에 1~3자리까지 조건에 적합하지만 문자열 길이가 4인 경우

icraft2170 committed:
etc : #6 기록용 커밋
- 단위 테스트 작성 후 로직 수정해 봄.
- 테스트 이후 조건에 맞지 않은 경우가 발생함.
- 테스트 케이스별로 단위테스트를 작성하는 것이 왜 중요한지 예시가 될 수 있는 상황

icraft2170 committed:
fix : #6 입력된 문자열 검증 로직에 논리적 오류 수정

icraft2170 commented:
@ahngj96 @f-lab-elkein 

유닛테스트 및 로직 수정했습니다!! 🙏🙏🙏 감사합니다.
확인 끝나시면, 머지 해주시거나 댓글 남겨주시면 바로 제가 머지 하도록 하겠습니다.

ahngj96 commented:
루프 안에서 break나 return 사용을 지양하시나 보군요 ㅎㅎ

확인했습니다! 👍
